### ISSUE TITLE
Parallel middleware

### ISSUE THREAD

**Issue by kgriffs** [falconry/falcon#65]
Add optional support for executing ingress hooks in parallel using greenlets, asyncio, and/or Twisted. Processing should not continue until all parallel hooks return success (need to be able to specify conditions on the ctx that must all be true before proceeding to the next group). If any hook fails/bails out, the remaining hooks and the main app do not execute, and any parallel hooks that are still running are killed.

**Comment by kgriffs** [comment]
Now that we have middleware, this functionality would better fit there. May make sense as a standalone library vs. being part of Falcon core.

**Comment by vytas7** [comment]
I have renamed the issue to parallel middleware, which could be implemented via [`asyncio.gather()`](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather) (or a similar `anyio` primitive if we ever tread that path). Although ostensibly straightforward, this might be a can of worms wrt synchronizing interaction with the `req` and `resp` objects, particularly when doing I/O; input/output messages can get interleaved between different components, etc.

In that light, maybe we should close this issue in favour of an add-on library (should anyone build this)? @kgriffs @CaselIT

**Comment by CaselIT** [comment]
I agree with your sentiment @vytas7
I don't think we need to implement this feature, since it's different compared to how middleware are generally used both in falcon and in other frameworks. I agree that sync may be a can of worms

**Comment by vytas7** [comment]
Closing in favour of implementing this as a separate add-on (or Gist).
